rbd/cache: Replicated Write Log core codes part 2

yison opened the pull request:
Adds Replicated Write Log, a persistent, write back cache for Ceph RBD.
Implements: http://tracker.ceph.com/projects/ceph/wiki/Rbd_-_ordered_crash-consistent_write-back_caching_extension

Trello: https://trello.com/c/QnsQaGTn

[Obsoletes https://github.com//pull/29087]

The PR 29087 is split into smaller PRs to make review easier.
The first PR: https://github.com/ceph/ceph/pull/31279

This is the second PR which defines the following parts:

add log entry class
add sync point class
add log operation classes
add WriteRequest to handle write
aio_write to flush data into cache device
add aio_write tests

Signed-off-by: Scott Peterson scott.d.peterson@intel.com
Signed-off-by: Lisa Li xiaoyan.li@intel.com
Signed-off-by: Yuan Lu yuan.y.lu@intel.com
Signed-off-by: Mahati Chamarthy mahati.chamarthy@intel.com

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +570,733 @@ template <typename I>
+
+  int r = m_write_log_guard.detain(req.block_extent, &req, &cell);
+  ceph_assert(r>=0);
+  if (r > 0) {
```
We check r with ceph_assert, so no need if here.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +570,733 @@ template <typename I>
+  for (auto &operation : ops) {
+    uint32_t entry_index = m_first_free_entry;
+    m_first_free_entry = (m_first_free_entry + 1) % m_total_log_entries;
+    //if (m_log_entries.back()) {
```
Need to remove the comments.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +570,733 @@ template <typename I>
+  {
+    std::lock_guard locker(m_lock);
+    m_unpublished_reserves -= published_reserves;
+    m_dirty_log_entries.splice(m_dirty_log_entries.end(), dirty_entries);
```
The dirty log entries include all the cache log entries in the cache device. Flushing cached data on a cache device to OSD will be handled in next PR.

lixiaoy1 commented:
This PR includes the flow of aio_write, and includes the handling of write requests and flushing data to a cache device. The flow to flushing cached data from the cache device to OSD is implemented in later PRs.   
It supports two persistent modes:   
1. persistent_on_write: per aio_write, it is persistent to the cache device before it returns (calls on_finish).  It adds a SyncPoint per every write request.
2. persistent_on_flush: per aio_write, write requests are persistent to the cache device only when a flush request is sent. The SyncPoint is appended per flush request. 

@dillaman We tried to make the PR smaller, but to make the functionality and logic intact we delivered this PR.  Anyway later PRs will be much smaller.  Let us know your suggestions.

dillaman commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+ * may be completed from here before the IO persists.
+ */
+template <typename T>
+class C_BlockIORequest : public C_GuardedBlockIORequest<T> {
```
Nit: can these two classes be collapsed?

lixiaoy1 replied:
They can be combined.

dillaman commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+
+class GuardedRequestFunctionContext : public Context {
+public:
+  BlockGuardCell *m_cell = nullptr;
```
Nit: drop `m_` if these are going to be publicly available (here and below)

dillaman commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+  int fadvise_flags;
+  Context *user_req; /* User write request */
+  ExtentsSummary<io::Extents> image_extents_summary;
+  bool detained = false;                /* Detained in blockguard (overlapped with a prior IO) */
```
Any reason this is public?

lixiaoy1 replied:
Currently, it is updated directly in `ReplicatedWriteLog<I>::release_guarded_request()`, let's improve the Encapsulation.

dillaman commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+  virtual void update_req_stats(utime_t &now) {
+    // TODO: Add in later PRs
+  }
+  virtual bool alloc_resources() override;
```
Nit: drop `virtual` keyword (you have `override`)

lixiaoy1 replied:
OK.

dillaman commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+  virtual bool alloc_resources() override;
+
+  /* Plain writes will allocate one buffer per request extent */
+  virtual void setup_buffer_resources(uint64_t &bytes_cached, uint64_t &bytes_dirtied);
```
Nit: any reason why these few methods are made virtual?

lixiaoy1 replied:
We will have C_WriteSameRequest inherited from C_WriteRequest. It has its own `setup_buffer_resources`.

dillaman replied:
Looking at [1], it seems like it's just replacing the function? Perhaps this would be better served with an abstract base "write-op" base class?

lixiaoy1 replied:
Like the comment about alloc_resouces  https://github.com/ceph/ceph/pull/31963#discussion_r355243192, we need to have an abstract base write class.

lixiaoy1 replied:
Sorry, we don't need to create an abstract base write class. the behaviours of alloc_resource are same for all write requests, including normal write, writesame and compareandwrite.

dillaman commented on src/librbd/cache/rwl/Request.cc:
```diff
@@ -0,0 +1,557 @@
+  resources.buffers.reserve(this->image_extents.size());
+  {
+    std::lock_guard locker(rwl.m_lock);
+    if (rwl.m_free_lanes < this->image_extents.size()) {
```
General delegation of responsibility comment: this class could ask the RWL for the required space and get the references to the spaces back from the RWL -- instead of embedding this logic in each request class.

lixiaoy1 replied:
The check is only for write requests which need to allocate data space in the pmem device. It controls the number of concurrently asking for space allocation on the pmem device. C_WriteRequest is the class for normal write requests, also it is the base class of  C_CompAndWriteRequest and C_WriteSameRequest.   So the check is not redundant in other places.

lixiaoy1 replied:
We would like to move alloc_resources to the ReplicatedWriteLog class and overloads it with different Requests. eg,
```c++
bool alloc_resources(C_CompAndWriteRequest *req);
bool alloc_resources(C_WriteSameRequest *req);
...
```

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +570,733 @@ template <typename I>
+      if (RWL_VERBOSE_LOGGING) {
+        ldout(m_image_ctx.cct, 20) << "deferred IOs: " << m_deferred_ios.size() << dendl;
+      }
+      dispatch_deferred_writes();
```
Nit: should this really expect to be able to dispatch anything here? or should deferred writes be dispatched upon IO completions to the pmem device/flush to the OSD?

lixiaoy1 replied:
I agree with you in the theory. This should be called when some resource is freed. Let us do more check.

lixiaoy1 replied:
We have to call it. For example, when we checking m_deferred_ios, it is not empty. After the check in Line 771 and before it is put into deferred_ios, dispatch_deferred_writes is triggered and handles all the deferred ios. At this time the resource may be enough but not trigger the dispatch.

dillaman replied:
Line 771?

lixiaoy1 replied:
I mean these lines https://github.com/ceph/ceph/pull/29078/files#diff-c01327ef02947ca979424bf5823abc22R769-R770

lixiaoy1 replied:
Meanwhile, it is the similar situation when alloca_resources fails for a write request. We don't get ReplicatedWriteLog m_lock all the time. It is possible that the allocation fails but new resources are released before the write request is added into the m_deferred_ios queue. https://github.com/ceph/ceph/blob/edb10684a5521e94ad707fa9289e12ceb33d8459/src/librbd/cache/ReplicatedWriteLog.cc#L773

lixiaoy1 commented on src/librbd/cache/rwl/Types.h:
```diff
@@ -135,10 +137,27 @@ enum {
   l_librbd_rwl_last,
 };
 
+const uint32_t MIN_WRITE_ALLOC_SIZE = 512;
```
OK

lixiaoy1 commented on src/librbd/CMakeLists.txt:
```diff
@@ -152,6 +152,9 @@ if(WITH_RBD_RWL)
     cache/rwl/Types.cc
     cache/rwl/LogEntry.cc
     cache/rwl/ImageCacheState.cc
+    cache/rwl/SyncPoint.cc
```
Will do it

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -25,32 +25,32 @@
+typedef ReplicatedWriteLog<ImageCtx>::Extent Extent;
 typedef ReplicatedWriteLog<ImageCtx>::Extents Extents;
 
+const unsigned long int ops_appended_together = MAX_ALLOC_PER_TRANSACTION;
```
OK

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -25,32 +25,32 @@
     m_log_pool_config_size(DEFAULT_POOL_SIZE),
-    m_image_writeback(image_ctx),
+    m_image_writeback(image_ctx), m_write_log_guard(image_ctx.cct),
+    m_deferred_dispatch_lock("librbd::cache::ReplicatedWriteLog::m_deferred_dispatch_lock",
```
m_deferred_dispatch_lock is used to make sure that only one thread is dispatching deferred ios and deferred ios are dispatched in order.  The usage is in `ReplicatedWriteLog<I>::dispatch_deferred_writes()`.   The deferred ios (`m_deferred_ios`) are the ios which are waiting for resources.     

m_blockguard_lock is to handle the competition of write barrier.  All the IOs go through m_write_log_guard sequentially. Please refer to `ReplicatedWriteLog<I>::detain_guarded_request`.     

m_log_append_lock is to handle the competition of updating pool_root and appending logs to the cache device.

lixiaoy1 replied:
About m_entry_bl_block, it is handled the competition of init_pmem_bl(). It should be just related to WriteLogEntry, we will confirm and move it to WriteLogEntry inside.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -417,6 +417,9 @@ void ReplicatedWriteLog<I>::rwl_init(Context *on_finish, DeferredContexts &later
   // TODO: Will init sync point, this will be covered in later PR.
   //  init_flush_new_sync_point(later);
+  ++m_current_sync_gen;
+  auto new_sync_point = std::make_shared<SyncPointT>(*this, m_current_sync_gen);
```
We will remove it

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -479,6 +496,41 @@ void ReplicatedWriteLog<I>::aio_write(Extents &&image_extents,
                                       int fadvise_flags,
                                       Context *on_finish) {
+  CephContext *cct = m_image_ctx.cct;
+  if (RWL_VERBOSE_LOGGING) {
```
We will remove RWL_VERBOSE_LOGGING.

dillaman replied:
You can also create a new `rbd_rwl` subsystem if you just want to enable logging within RWL independently from rbd.

lixiaoy1 replied:
This seems better, we will create a new rbd_rwl log subsystem.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -479,6 +496,41 @@ void ReplicatedWriteLog<I>::aio_write(Extents &&image_extents,
+  m_perfcounter->inc(l_librbd_rwl_wr_req, 1);
+
+  ceph_assert(m_initialized);
+  {
```
Will remove Line 507 - 513.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -479,6 +496,41 @@ void ReplicatedWriteLog<I>::aio_write(Extents &&image_extents,
+    }
+  }
+
+  if (ExtentsSummary<Extents>(image_extents).total_bytes == 0) {
```
I will submit another PR to handle.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -479,6 +496,41 @@ void ReplicatedWriteLog<I>::aio_write(Extents &&image_extents,
+      alloc_and_dispatch_io_req(write_req);
+    });
+
+  detain_guarded_request(GuardedRequest(write_req->image_extents_summary.block_extent(),
```
OK, we will give C_BlockIORequest pointer to `detain_guarded_request()`. This decrease the code redundancy of grabbing.

lixiaoy1 replied:
Yes, we can give a `C_BlockIORequest` pointer. @yison in later PR we need to add a parameter called is_barrier for flush request.

yison replied:
C_BlockIORequest a template, not a class. If we use  C_BlockIORequest pointer, we need to make GuardedRequest become a template too.  But I don't think it's a good idea to make GuardedRequest  become a template due to C_BlockIORequest pointer.

lixiaoy1 replied:
We can give C_BlockIORequest  as the parameter of detain_guarded_request(), and create GuardedRequest inside so that the codes will not be redundant.

dillaman replied:
It would be interesting to know, given my other comments about proper encapsulation, if those really need to be templates. For example, could they just take a pointer to a simple interface for all the hooks they need back into RWL?

lixiaoy1 replied:
Sorry @dillaman, I don't quite understand your concern. In fact, I just had a check. If we transfer the pointer of C_BlockIORequest to GuardedRequest's constructor, the constructor needs to be a template function and the GuardedRequest doesn't need to be a template class. 
```cpp
class GuardedRequest {
  ....
  template<typename T> GuardedRequest(C_BlockIORequest<T> *p) {
    ...
  }
}
```

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +570,733 @@ template <typename I>
 } // namespace cache
 } // namespace librbd
 
+#ifndef TEST_F
```
During unit test, no need to Instantiate these template classes. Instead it instantiates with MockImageCtx.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.h:
```diff
@@ -68,6 +94,35 @@ class ReplicatedWriteLog : public ImageCache<ImageCtxT> {
+  using C_BlockIORequestT = rwl::C_BlockIORequest<This>;
+  using C_WriteRequestT = rwl::C_WriteRequest<This>;
+
+  friend class rwl::SyncPoint<This>;
```
We will improve it.

lixiaoy1 commented on src/librbd/cache/rwl/LogEntry.h:
```diff
@@ -28,11 +34,135 @@ class GenericLogEntry {
+                                  const SyncPointLogEntry &entry);
+};
+
+class GeneralWriteLogEntry : public GenericLogEntry {
```
let's update it to GenericWriteLogEntry.

lixiaoy1 commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+                                  const BlockGuardReqState &r);
+};
+
+class GuardedRequestFunctionContext : public Context {
```
We can't as we transfer *this to the callback.  
```c++
void GuardedRequestFunctionContext::finish(int r) {
  assert(m_cell);
  m_callback(*this);
}```

While in LambContext, 
```c++
  void finish(int r) override {
    t(r);
  }
```

lixiaoy1 commented on src/librbd/cache/rwl/Request.h:
```diff
@@ -0,0 +1,200 @@
+  BlockGuardCell *m_cell = nullptr;
+  BlockGuardReqState m_state;
+  GuardedRequestFunctionContext(boost::function<void(GuardedRequestFunctionContext&)> &&callback);
+  ~GuardedRequestFunctionContext(void);
```
OK.

lixiaoy1 commented on src/librbd/cache/rwl/SyncPoint.h:
```diff
@@ -0,0 +1,64 @@
+namespace cache {
+namespace rwl {
+/* Limit work between sync points */
+static const uint64_t MAX_WRITES_PER_SYNC_POINT = 256;
```
As it is not only used in SyncPoint.cc, we need to put it in a header file. It is not very reasonable to put it in SyncPoint. How about we put it in rwl/Types.h?

yison replied:
I decide to move this to rwl/Types.h.  We'll also use this in other classes, not only SyncPoint class.

lixiaoy1 commented on src/librbd/cache/rwl/SyncPoint.h:
```diff
@@ -0,0 +1,64 @@
+/* Limit work between sync points */
+static const uint64_t MAX_WRITES_PER_SYNC_POINT = 256;
+
+template <typename T>
```
Agree with you.

lixiaoy1 commented on src/librbd/cache/rwl/SyncPoint.h:
```diff
@@ -0,0 +1,64 @@
+  std::ostream &format(std::ostream &os) const;
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const SyncPoint &p) {
+    return p.format(os);
```
There is no value in SyncPoint. Here it just wants to keep the same style as LogEntry related classes. Will remove it.

lixiaoy1 commented on src/librbd/cache/rwl/Types.cc:
```diff
@@ -22,7 +22,27 @@ DeferredContexts::~DeferredContexts() {
+ * convert between image and block extents here using a "block size"
+ * of 1.
+ */
+const BlockExtent block_extent(const uint64_t offset_bytes, const uint64_t length_bytes)
```
OK.

lixiaoy1 commented on src/librbd/cache/rwl/Types.cc:
```diff
@@ -22,7 +22,27 @@ DeferredContexts::~DeferredContexts() {
+}
+
+const BlockExtent WriteLogPmemEntry::block_extent() {
+  return BlockExtent(librbd::cache::rwl::block_extent(image_offset_bytes, write_bytes));
```
OK.

lixiaoy1 commented on src/librbd/cache/rwl/Types.cc:
```diff
@@ -72,6 +92,40 @@ std::ostream& operator<<(std::ostream& os,
 
+template <typename ExtentsType>
+ExtentsSummary<ExtentsType>::ExtentsSummary(const ExtentsType &extents) {
+  total_bytes = 0;
```
OK.

lixiaoy1 commented on src/librbd/cache/rwl/Types.h:
```diff
@@ -208,6 +228,36 @@ struct WriteLogPoolRoot {
+  uint64_t total_bytes;
+  uint64_t first_image_byte;
+  uint64_t last_image_byte;
+  ExtentsSummary(const ExtentsType &extents);
```
OK.

lixiaoy1 commented on src/librbd/cache/rwl/Types.h:
```diff
@@ -178,6 +197,7 @@ struct WriteLogPmemEntry {
     : image_offset_bytes(image_offset_bytes), write_bytes(write_bytes),
       entry_valid(0), sync_point(0), sequenced(0), has_data(0), discard(0), writesame(0) {
   }
+  const BlockExtent block_extent();
```
Agree

lixiaoy1 commented on src/librbd/cache/rwl/Types.h:
```diff
@@ -208,6 +228,36 @@ struct WriteLogPoolRoot {
+  const BlockExtent block_extent() {
+    return BlockExtent(first_image_byte, last_image_byte);
+  }
+  const io::Extent image_extent(const BlockExtent& block_extent) {
```
Agree.

yison commented:
@dillaman Hi Jason, we modified codes based on your suggestions. Please review it when you're convenient. Thanks a lot.

yison commented:
@dillaman Hi, we did changes based on your comments. Please review it when you're convenient. Thanks a lot.

dillaman commented:
Just to keep you in the loop, I plan to dig into this tomorrow morning as my first priority.

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +554,889 @@ template <typename I>
+ * Lanes are released after the write persists via release_write_lanes()
+ */
+template <typename I>
+bool ReplicatedWriteLog<I>::alloc_resources(C_WriteRequestT *req) {
```
Nit: I would still like to see a generic core "allocate_resources" method that can take a generic `C_BlockIORequest` and query the request for the number of buffers and sizes required for the op. Alternatively, flip it around so that `C_BlockIORequest` just asks RWL for X resources generically.

From a quick look/suggestion: have the write request just call a RWL generic allocate resource method where it passes the pre-setup `resources->buffers`. The RWL can use those to perform the checks for lanes, log entries, bytes allocated, etc.

lixiaoy1 replied:
Currently common part is here, and getting required resource is special in every request.

lixiaoy1 replied:
Currently common part is here, and getting required resource is special in every request.

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +554,889 @@ template <typename I>
+    m_perfcounter->tinc(l_librbd_rwl_log_op_dis_to_cmp_t, now - op->dispatch_time);
+    m_perfcounter->hinc(l_librbd_rwl_log_op_dis_to_cmp_t_hist, utime_t(now - op->dispatch_time).to_nsec(),
+                      log_entry->ram_entry.write_bytes);
+    if (op->is_write()) {
```
Nit: you already have a `complete` method on the log operation, so just pass `m_perfcounter` to it so that the derived log operation can directly update its stats.

lixiaoy1 replied:
DONE

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +554,889 @@ template <typename I>
+    if (op->is_write() || op->is_writesame()) {
+      published_reserves++;
+    }
+    if (op->is_discard()) {
```
Nit: just log this directly in the derived class.

lixiaoy1 replied:
DONE

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +554,889 @@ template <typename I>
+      op->get_gen_write_op()->sync_point->log_entry->writes_completed++;
+      dirty_entries.push_back(log_entry);
+    }
+    if (op->is_write() || op->is_writesame()) {
```
Nit: instead of `is_write` and `is_writesame` checks, have a `reserved_allocated`-like method

lixiaoy1 replied:
DONE

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +554,889 @@ template <typename I>
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::setup_log_operations(C_WriteRequestT &write_req, utime_t dispatched_time) {
```
This function is moved to request itself.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +554,889 @@ template <typename I>
+}
+
+template <typename I>
+bool ReplicatedWriteLog<I>::append_write_request(C_WriteRequestT *write_req, std::shared_ptr<SyncPoint> sync_point) {
```
This function is moved to C_WriteRequest which uses it.

lixiaoy1 commented:
@dillaman Thank you for your review. And your comments are addressed, please have a look.

dillaman commented on src/librbd/cache/rwl/Request.cc:
```diff
@@ -0,0 +1,369 @@
+    bl(std::move(bl)), fadvise_flags(fadvise_flags),
+    user_req(user_req), image_extents_summary(image_extents), m_arrived_time(arrived) {
+  ldout(rwl.get_context(), 99) << this << dendl;
+  /* Remove zero length image extents from input */
```
Nit: this is handled at a higher level now, right?

lixiaoy1 replied:
I added a change here https://github.com/ceph/ceph/blob/master/src/librbd/io/ImageRequest.cc#L424-L430 to check whether the total length is 0.  
I also think the handler should be at a higher level. But from the existed codes, I didn't find a case that there are more than 1 extents in an ImageRequest. Do you think we need to add the handler?

dillaman replied:
I *believe* support is only there to handle reads from parent images -- the write path is unused at the moment.

lixiaoy1 replied:
C_BlockIORequest is a parent class for write/flush/discard/writesame/compareandwrite requests. And it is not a parent class of read request. So this problem should not exist.

dillaman commented on src/librbd/cache/rwl/Types.h:
```diff
@@ -178,13 +198,14 @@ struct WriteLogPmemEntry {
   bool is_sync_point();
   bool is_discard();
   bool is_writesame();
   bool is_write();
```
Nit: are these used? Can they be removed and instead just rely on polymorphic methods to handle any deltas in behaviour.

lixiaoy1 replied:
I removed the interfaces. Currently, they are not used.

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,203 @@
+  std::ostream &format(std::ostream &os) const;
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const SyncPointLogOperation &op);
+  const std::shared_ptr<GenericLogEntry> get_log_entry() {
```
Nit: this applies for remainder of methods here (and the virtual methods in other derived classes), but make sure you append `override` suffix on virtual method overrides

lixiaoy1 replied:
Set it override.

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +555,778 @@ template <typename I>
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::detain_guarded_request(C_BlockIORequestT *request, GuardedRequestFunctionContext *guarded_ctx)
```
Nit: wrap down to next line at around 80 characters

lixiaoy1 replied:
Done

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +555,778 @@ template <typename I>
+        std::advance(last_in_batch, ops_to_append);
+        ops.splice(ops.end(), m_ops_to_append, m_ops_to_append.begin(), last_in_batch);
+        ops_remain = true; /* Always check again before leaving */
+        ldout(m_image_ctx.cct, 20) << "appending " << ops.size() << ", " << m_ops_to_append.size() << " remain" << dendl;
```
Nit: wrap down to next line

lixiaoy1 replied:
Done

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +555,778 @@ template <typename I>
+        std::advance(last_in_batch, ops_to_flush);
+        ops.splice(ops.end(), m_ops_to_flush, m_ops_to_flush.begin(), last_in_batch);
+        ops_remain = !m_ops_to_flush.empty();
+        ldout(m_image_ctx.cct, 20) << "flushing " << ops.size() << ", " << m_ops_to_flush.size() << " remain" << dendl;
```
Nit: wrap down to next line

lixiaoy1 replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,203 @@
+  }
+  virtual void appending() = 0;
+  virtual void complete(int r) = 0;
+  virtual bool is_write() {
```
Nit: sorry -- same comment here re: removal of these IO methods if possible

lixiaoy1 replied:
I deleted such interfaces.

lixiaoy1 commented:
@dillaman Thank you for the review. The PR is updated, please have a look.

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+  const std::shared_ptr<GenericLogEntry> get_log_entry() override {
+    return sync_point->log_entry;
+  }
+  void appending();
```
Nit: please make sure all overridden methods are properly tagged w/ the `override` keyword (this comment applies throughout).

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+    return true;
+  }
+  void appending();
+  virtual void complete(int r);
```
Nit: the `virtual` keyword should only be on the lowest-level of the virtual function. This should have the `override` keyword.

yison replied:
Done

dillaman commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -513,9 +559,789 @@ template <typename I>
+    auto log_entry = op->get_log_entry();
+    log_entry->completed = true;
+    if (op->reserved_allocated()) {
+      op->get_gen_write_op()->sync_point->log_entry->writes_completed++;
```
Nit: drop the `get_gen_write_op` virtual method and instead create something like `mark_log_entry_completed()`

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+  }
+  virtual void appending() = 0;
+  virtual void complete(int r) = 0;
+  virtual GenericWriteLogOperation *get_gen_write_op() {
```
Nit: drop this method and replace w/ a generic method (to increase the metrics in this case)

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+  std::ostream &format(std::ostream &os) const;
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const GenericWriteLogOperation &op);
+  GenericWriteLogOperation *get_gen_write_op() {
```
Nit: drop

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/Request.cc:
```diff
@@ -0,0 +1,361 @@
+    uint64_t buffer_offset = 0;
+    for (auto &gen_op : op_set->operations) {
+      /* A WS is also a write */
+      auto operation = gen_op->get_write_op();
```
Nit: drop `get_write_op` and replace w/ a generic handler method that can be properly overridden in the write op derived class. Alternatively, a few lines prior you created a `WriteLogOperationSet`, so just retain the derived pointer so that you have direct access to the necessary write ops w/o the need for casting nor magic helper methods.

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/Request.cc:
```diff
@@ -0,0 +1,361 @@
+  /* Write data */
+  for (auto &operation : op_set->operations) {
+    /* operation is a shared_ptr, so write_op is only good as long as operation is in scope */
+    auto write_op = operation->get_write_op();
```
Nit: drop `get_write_op` and replace w/ a generic handler method that can be properly overridden in the write op derived class.

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+  virtual GenericWriteLogOperation *get_gen_write_op() {
+    return nullptr;
+  }
+  virtual WriteLogOperation *get_write_op() {
```
Nit: drop this method

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+  const std::shared_ptr<WriteLogEntry> get_write_log_entry() override {
+    return log_entry;
+  }
+  WriteLogOperation *get_write_op() override {
```
Nit: drop this method

yison replied:
Done

dillaman commented on src/librbd/cache/rwl/LogOperation.h:
```diff
@@ -0,0 +1,183 @@
+  const std::shared_ptr<GenericLogEntry> get_log_entry() override {
+    return get_write_log_entry();
+  }
+  const std::shared_ptr<WriteLogEntry> get_write_log_entry() override {
```
Nit: drop this method

yison replied:
Done

yison commented:
@dillaman Thank you for the review. The PR is updated, please have a look.

dillaman commented:
jenkins test make check

dillaman commented:
This actually fails to compile for me under GCC 9 due to this missing PR in pmdk: https://github.com/pmem/pmdk/pull/3732. Should the pmdk library be updated in Ceph?

dillaman commented:
Lots of compile errors after manually patching pmdk for GCC 9:

> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(std::shared_ptr<librbd::cache::rwl::SyncPointLogEntry>, uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:120:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   120 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:125:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   125 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/common/dout.h:21,
>                  from /home/jdillaman/ceph_wip/src/librbd/BlockGuard.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Types.h:9,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc: In member function ‘void librbd::cache::rwl::WriteLogEntry::init_pmem_bp()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:90:24: error: ‘class ceph::buffer::v14_2_0::ptr’ has no member named ‘get_raw’; did you mean ‘ceph::buffer::v14_2_0::raw* ceph::buffer::v14_2_0::ptr::_raw’? (not accessible from this context)
>    90 |   ceph_assert(!pmem_bp.get_raw());
>       |                        ^~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> In file included from /home/jdillaman/ceph_wip/src/include/encoding.h:38,
>                  from /home/jdillaman/ceph_wip/src/include/uuid.h:9,
>                  from /home/jdillaman/ceph_wip/src/include/types.h:21,
>                  from /home/jdillaman/ceph_wip/src/msg/msg_types.h:23,
>                  from /home/jdillaman/ceph_wip/src/common/options.h:11,
>                  from /home/jdillaman/ceph_wip/src/common/config.h:23,
>                  from /home/jdillaman/ceph_wip/src/common/config_proxy.h:6,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_context.h:36,
>                  from /home/jdillaman/ceph_wip/src/common/dout.h:28,
>                  from /home/jdillaman/ceph_wip/src/librbd/BlockGuard.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Types.h:9,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/include/buffer.h:187:10: note: declared protected here
>   187 |     raw *_raw;
>       |          ^~~~
> In file included from /home/jdillaman/ceph_wip/src/common/dout.h:21,
>                  from /home/jdillaman/ceph_wip/src/librbd/BlockGuard.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Types.h:9,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc: In member function ‘void librbd::cache::rwl::WriteLogEntry::init_pmem_bl()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:97:23: error: ‘class ceph::buffer::v14_2_0::ptr’ has no member named ‘get_raw’; did you mean ‘ceph::buffer::v14_2_0::raw* ceph::buffer::v14_2_0::ptr::_raw’? (not accessible from this context)
>    97 |   ceph_assert(pmem_bp.get_raw());
>       |                       ^~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> In file included from /home/jdillaman/ceph_wip/src/include/encoding.h:38,
>                  from /home/jdillaman/ceph_wip/src/include/uuid.h:9,
>                  from /home/jdillaman/ceph_wip/src/include/types.h:21,
>                  from /home/jdillaman/ceph_wip/src/msg/msg_types.h:23,
>                  from /home/jdillaman/ceph_wip/src/common/options.h:11,
>                  from /home/jdillaman/ceph_wip/src/common/config.h:23,
>                  from /home/jdillaman/ceph_wip/src/common/config_proxy.h:6,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_context.h:36,
>                  from /home/jdillaman/ceph_wip/src/common/dout.h:28,
>                  from /home/jdillaman/ceph_wip/src/librbd/BlockGuard.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Types.h:9,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/include/buffer.h:187:10: note: declared protected here
>   187 |     raw *_raw;
>       |          ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc: In member function ‘unsigned int librbd::cache::rwl::WriteLogEntry::reader_count()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:105:15: error: ‘class ceph::buffer::v14_2_0::ptr’ has no member named ‘get_raw’; did you mean ‘ceph::buffer::v14_2_0::raw* ceph::buffer::v14_2_0::ptr::_raw’? (not accessible from this context)
>   105 |   if (pmem_bp.get_raw()) {
>       |               ^~~~~~~
> In file included from /home/jdillaman/ceph_wip/src/include/encoding.h:38,
>                  from /home/jdillaman/ceph_wip/src/include/uuid.h:9,
>                  from /home/jdillaman/ceph_wip/src/include/types.h:21,
>                  from /home/jdillaman/ceph_wip/src/msg/msg_types.h:23,
>                  from /home/jdillaman/ceph_wip/src/common/options.h:11,
>                  from /home/jdillaman/ceph_wip/src/common/config.h:23,
>                  from /home/jdillaman/ceph_wip/src/common/config_proxy.h:6,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_context.h:36,
>                  from /home/jdillaman/ceph_wip/src/common/dout.h:28,
>                  from /home/jdillaman/ceph_wip/src/librbd/BlockGuard.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Types.h:9,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/include/buffer.h:187:10: note: declared protected here
>   187 |     raw *_raw;
>       |          ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:110:1: warning: control reaches end of non-void function [-Wreturn-type]
>   110 | }
>       | ^
> make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/build.make:1974: src/librbd/CMakeFiles/rbd_internal.dir/cache/rwl/LogEntry.cc.o] Error 1
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.h:10,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.cc:4:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(std::shared_ptr<librbd::cache::rwl::SyncPointLogEntry>, uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:120:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   120 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/mutex:43,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_context.h:21,
>                  from /home/jdillaman/ceph_wip/src/common/dout.h:28,
>                  from /home/jdillaman/ceph_wip/src/include/Context.h:19,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.cc:4:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.h:10,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.cc:4:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:125:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   125 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/mutex:43,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_context.h:21,
>                  from /home/jdillaman/ceph_wip/src/common/dout.h:28,
>                  from /home/jdillaman/ceph_wip/src/include/Context.h:19,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/Request.cc:4:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/build.make:2000: src/librbd/CMakeFiles/rbd_internal.dir/cache/rwl/Request.cc.o] Error 1
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:15,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(std::shared_ptr<librbd::cache::rwl::SyncPointLogEntry>, uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:120:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   120 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/common/WorkQueue.h:33,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:15,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:125:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   125 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/common/WorkQueue.h:33,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/common/RWLock.h:22,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In member function ‘void librbd::cache::ReplicatedWriteLog<ImageCtxT>::arm_periodic_stats()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:288:29: error: ‘ceph::mutex’ {aka ‘class std::mutex’} has no member named ‘is_locked’
>   288 |   ceph_assert(m_timer_lock->is_locked());
>       |                             ^~~~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In member function ‘librbd::BlockGuardCell* librbd::cache::ReplicatedWriteLog<ImageCtxT>::detain_guarded_request_helper(librbd::cache::rwl::GuardedRequest&)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:573:33: error: ‘ceph::mutex’ {aka ‘class std::mutex’} has no member named ‘is_locked_by_me’
>   573 |   ceph_assert(m_blockguard_lock.is_locked_by_me());
>       |                                 ^~~~~~~~~~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In member function ‘librbd::BlockGuardCell* librbd::cache::ReplicatedWriteLog<ImageCtxT>::detain_guarded_request_barrier_helper(librbd::cache::rwl::GuardedRequest&)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:594:33: error: ‘ceph::mutex’ {aka ‘class std::mutex’} has no member named ‘is_locked_by_me’
>   594 |   ceph_assert(m_blockguard_lock.is_locked_by_me());
>       |                                 ^~~~~~~~~~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In member function ‘void librbd::cache::ReplicatedWriteLog<ImageCtxT>::alloc_op_log_entries(librbd::cache::rwl::GenericLogOperations&)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:910:33: error: ‘ceph::mutex’ {aka ‘class std::mutex’} has no member named ‘is_locked_by_me’
>   910 |   ceph_assert(m_log_append_lock.is_locked_by_me());
>       |                                 ^~~~~~~~~~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In member function ‘int librbd::cache::ReplicatedWriteLog<ImageCtxT>::append_op_log_entries(librbd::cache::rwl::GenericLogOperations&)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:968:33: error: ‘ceph::mutex’ {aka ‘class std::mutex’} has no member named ‘is_locked_by_me’
>   968 |   ceph_assert(m_log_append_lock.is_locked_by_me());
>       |                                 ^~~~~~~~~~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In member function ‘bool librbd::cache::ReplicatedWriteLog<ImageCtxT>::alloc_resources(librbd::cache::ReplicatedWriteLog<ImageCtxT>::C_BlockIORequestT*)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:1233:23: error: ‘ceph::mutex’ {aka ‘class std::mutex’} has no member named ‘is_locked_by_me’
>  1233 |   ceph_assert(!m_lock.is_locked_by_me());
>       |                       ^~~~~~~~~~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc: In instantiation of ‘librbd::cache::ReplicatedWriteLog<ImageCtxT>::ReplicatedWriteLog(I&, librbd::cache::rwl::ImageCacheState<ImageCtxT>*) [with ImageCtxT = librbd::ImageCtx]’:
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:1341:31:   required from here
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:58:32: error: no matching function for call to ‘std::mutex::mutex(const char [60], bool, bool)’
>    58 |                  &m_thread_pool)
>       |                                ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/common/WorkQueue.h:33,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:58:32: error: no matching function for call to ‘std::mutex::mutex(const char [53], bool, bool)’
>    58 |                  &m_thread_pool)
>       |                                ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/common/WorkQueue.h:33,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:58:32: error: no matching function for call to ‘std::mutex::mutex(const char [40], bool, bool)’
>    58 |                  &m_thread_pool)
>       |                                ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/common/WorkQueue.h:33,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:58:32: error: no matching function for call to ‘std::mutex::mutex(const char [53], bool, bool)’
>    58 |                  &m_thread_pool)
>       |                                ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/common/WorkQueue.h:33,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/ReplicatedWriteLog.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/build.make:2039: src/librbd/CMakeFiles/rbd_internal.dir/cache/ReplicatedWriteLog.cc.o] Error 1
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.cc:4:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(std::shared_ptr<librbd::cache::rwl::SyncPointLogEntry>, uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:120:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   120 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/librbd/ImageCtx.h:16,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.cc:4:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.cc:4:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:125:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   125 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/condition_variable:39,
>                  from /home/jdillaman/ceph_wip/src/common/ceph_mutex.h:116,
>                  from /home/jdillaman/ceph_wip/src/librbd/ImageCtx.h:16,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/SyncPoint.cc:4:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/build.make:2013: src/librbd/CMakeFiles/rbd_internal.dir/cache/rwl/SyncPoint.cc.o] Error 1
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(std::shared_ptr<librbd::cache::rwl::SyncPointLogEntry>, uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:120:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   120 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/mutex:43,
>                  from /home/jdillaman/ceph_wip/src/common/convenience.h:15,
>                  from /home/jdillaman/ceph_wip/src/include/encoding.h:35,
>                  from /home/jdillaman/ceph_wip/src/include/uuid.h:9,
>                  from /home/jdillaman/ceph_wip/src/include/types.h:21,
>                  from /home/jdillaman/ceph_wip/src/include/utime.h:27,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> In file included from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h: In constructor ‘librbd::cache::rwl::WriteLogEntry::WriteLogEntry(uint64_t, uint64_t)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:125:33: error: no matching function for call to ‘std::mutex::mutex(const char [51], bool, bool)’
>   125 |                       true, true)
>       |                                 ^
> In file included from /usr/include/c++/9/mutex:43,
>                  from /home/jdillaman/ceph_wip/src/common/convenience.h:15,
>                  from /home/jdillaman/ceph_wip/src/include/encoding.h:35,
>                  from /home/jdillaman/ceph_wip/src/include/uuid.h:9,
>                  from /home/jdillaman/ceph_wip/src/include/types.h:21,
>                  from /home/jdillaman/ceph_wip/src/include/utime.h:27,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 3 provided
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc: In constructor ‘librbd::cache::rwl::GenericWriteLogOperation::GenericWriteLogOperation(std::shared_ptr<librbd::cache::rwl::SyncPoint>, utime_t, PerfCounters*, CephContext*)’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc:118:24: error: no matching function for call to ‘std::mutex::mutex(const char [53])’
>   118 |   sync_point(sync_point) {
>       |                        ^
> In file included from /usr/include/c++/9/mutex:43,
>                  from /home/jdillaman/ceph_wip/src/common/convenience.h:15,
>                  from /home/jdillaman/ceph_wip/src/include/encoding.h:35,
>                  from /home/jdillaman/ceph_wip/src/include/uuid.h:9,
>                  from /home/jdillaman/ceph_wip/src/include/types.h:21,
>                  from /home/jdillaman/ceph_wip/src/include/utime.h:27,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogOperation.cc:5:
> /usr/include/c++/9/bits/std_mutex.h:91:5: note: candidate: ‘constexpr std::mutex::mutex()’
>    91 |     mutex() noexcept = default;
>       |     ^~~~~
> /usr/include/c++/9/bits/std_mutex.h:91:5: note:   candidate expects 0 arguments, 1 provided
>

lixiaoy1 commented:
@dillaman Hi Jason, the gcc9 build issue is fixed in PMDK1.6.1, and this PR is going to update PMDK.

yison commented:
@dillaman Hi Jason, I have tried gcc9 build with PMDK tag 1.7. in my local machine. It built passed. Librbd unittests ran passed too. I just found that this PR:https://github.com/ceph/ceph/pull/32693 also used PMDK tag 1.7 and merged just now. You can have a try.

dillaman commented:
@yison The compile errors listed above still need to be fixed.

yison commented:
@dillaman I mean I have tested Replicated Write Log with gcc9 compile after using PMDK1.7 tag. It had no compile error. Did you still have compile error?

dillaman commented:
@yison I can only assume so since it was in RWL for the locks and had nothing to do w/ PMDK. I compiled in release mode which has different lock classes as compared to debug mode.

yison commented:
@dillaman Ok. I'll try release mode and have a look.

yison commented:
@dillaman Hi Jason, I fixed the compile issue in release mode and pushed again. Please have a look.

dillaman commented:
Looks like it fails to compile against `master`:

> In file included from /home/jdillaman/ceph_wip/src/include/rados/buffer.h:58,
>                  from /home/jdillaman/ceph_wip/src/include/rados/librados.hpp:11,
>                  from /home/jdillaman/ceph_wip/src/librbd/Utils.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc: In member function ‘void librbd::cache::rwl::WriteLogEntry::init_pmem_bp()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:90:24: error: ‘class ceph::buffer::v14_2_0::ptr’ has no member named ‘get_raw’; did you mean ‘ceph::buffer::v14_2_0::raw* ceph::buffer::v14_2_0::ptr::_raw’? (not accessible from this context)
>    90 |   ceph_assert(!pmem_bp.get_raw());
>       |                        ^~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> In file included from /home/jdillaman/ceph_wip/src/include/rados/librados.hpp:11,
>                  from /home/jdillaman/ceph_wip/src/librbd/Utils.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/include/rados/buffer.h:187:10: note: declared protected here
>   187 |     raw *_raw;
>       |          ^~~~
> In file included from /home/jdillaman/ceph_wip/src/include/rados/buffer.h:58,
>                  from /home/jdillaman/ceph_wip/src/include/rados/librados.hpp:11,
>                  from /home/jdillaman/ceph_wip/src/librbd/Utils.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc: In member function ‘void librbd::cache::rwl::WriteLogEntry::init_pmem_bl()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:97:23: error: ‘class ceph::buffer::v14_2_0::ptr’ has no member named ‘get_raw’; did you mean ‘ceph::buffer::v14_2_0::raw* ceph::buffer::v14_2_0::ptr::_raw’? (not accessible from this context)
>    97 |   ceph_assert(pmem_bp.get_raw());
>       |                       ^~~~~~~
> /home/jdillaman/ceph_wip/src/include/ceph_assert.h:106:6: note: in definition of macro ‘ceph_assert’
>   106 |    ((expr) \
>       |      ^~~~
> In file included from /home/jdillaman/ceph_wip/src/include/rados/librados.hpp:11,
>                  from /home/jdillaman/ceph_wip/src/librbd/Utils.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/include/rados/buffer.h:187:10: note: declared protected here
>   187 |     raw *_raw;
>       |          ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc: In member function ‘unsigned int librbd::cache::rwl::WriteLogEntry::reader_count()’:
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:105:15: error: ‘class ceph::buffer::v14_2_0::ptr’ has no member named ‘get_raw’; did you mean ‘ceph::buffer::v14_2_0::raw* ceph::buffer::v14_2_0::ptr::_raw’? (not accessible from this context)
>   105 |   if (pmem_bp.get_raw()) {
>       |               ^~~~~~~
> In file included from /home/jdillaman/ceph_wip/src/include/rados/librados.hpp:11,
>                  from /home/jdillaman/ceph_wip/src/librbd/Utils.h:7,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.h:8,
>                  from /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:5:
> /home/jdillaman/ceph_wip/src/include/rados/buffer.h:187:10: note: declared protected here
>   187 |     raw *_raw;
>       |          ^~~~
> /home/jdillaman/ceph_wip/src/librbd/cache/rwl/LogEntry.cc:110:1: warning: control reaches end of non-void function [-Wreturn-type]
>   110 | }
>       | ^
> make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/build.make:2013: src/librbd/CMakeFiles/rbd_internal.dir/cache/rwl/LogEntry.cc.o] Error 1
> make[3]: *** Waiting for unfinished jobs....
> make[2]: *** [CMakeFiles/Makefile2:21269: src/librbd/CMakeFiles/rbd_internal.dir/all] Error 2
> make[1]: *** [CMakeFiles/Makefile2:8844: src/tools/rbd/CMakeFiles/rbd.dir/rule] Error 2
> make: *** [Makefile:2437: rbd] Error 2

yison commented:
@dillaman Hi Jason, I rebased it with latest master branch and fixed the compile issue. (Both test in release mode and debug mode). Please have a look. Thanks.
